Category Service 구현

jintak0401 opened the pull request:
* Closes #33 
* Closes #34 
* Closes #45 

## ✨ **구현 기능 명세**
* [x] binary에서 char(6) 변경
* [x] Category Service 테스트 코드 구현
* [x] Category Service 기능 구현
    * [x] GET
    * [x] POST
    * [x] PUT
    * [x] DELETE

## 🎁 **주목할 점**

### mapToHexConverter
```ts
const strToHexColor = (color: string): THexColor => `#${color}`;

const mapToHexColor = <
  IType extends { color: string },
  TArgs = IType | IType[],
  TForm = { color: THexColor } & Omit<IType, 'color'>,
  TReturn = TArgs extends any[] ? TForm[] : TForm,
>(
  obj: TArgs,
): TReturn => {
  if (obj instanceof Array) {
    return obj.map(({ color, ...rest }) => ({
      ...rest,
      color: strToHexColor(color),
    })) as TReturn;
  }

  const { color, ...rest } = obj as unknown as IType;
  return { ...rest, color: strToHexColor(color) } as TReturn;
};
```
`mapToHexConverter`는 `{ color: string; }`을 포함하는 객체 혹은 객체의 배열을 인자를 받아(여기서 string은 /[a-zA-Z0-9]{6}$/ 를 만족하는 문자열입니다), 각 color의 맨 앞에 '#'을 붙인 형태로 return 해주는 함수입니다. `plan`과 `category`처럼 색상을 반환해주어야 하는 service가 여러개일 수 있으므로 제네릭을 이용하여 구현하였습니다. 

중간에 `const { color, ...rest } = obj as unknown as IType;` 가 있습니다. `as IType`이 아닌 `as unkown as IType`을 해준 이유는 다음과 같습니다.
1. `A as B`처럼 `as` 키워드를 사용할 수 있는 경우는 A가 B에 할당될 수 있을 경우, 혹은 B가 A에 할당될 수 있는 경우입니다.
2. obj의 타입은 `TArgs = IType | IType[]`입니다. 즉, `TArgs`는 `IType[]`이 될 수 있으므로 `IType[]`이 `IType`에 할당될 수 없으며, `IType`이 `IType[]`에 할당될 수 없습니다.
3. 따라서 어떤 것에도 `as`를 사용할 수 있게 해주는 `unkwnown`으로 변환해준 후, `IType`으로 변환해준 것입니다.

### Entity의 color 타입 변경

기존의 binary 타입은 몇몇 문제를 야기하므로 `char(6)` 타입으로 변경하였습니다. 

```ts
// Entity
  ...
  @ApiProperty()
  @Column({
    type: 'char',
    length: 6,
    default: '52d681',
  })
  @IsNotEmpty()
  @IsString()
  color!: string;
   ...

// DTO
  @ApiPropertyOptional({
    description: '카테고리 색상, #000000 ~ #ffffff',
    example: '#52d681',
  })
  @IsOptional()
  @IsHexColor()
  @Transform(({ value }) => value.replace('#', ''))
  color?: string;
```

### Jest setup

모든 테스트 코드에서 mocking 되어야 하는 것들을 미리 설정할 수 있는 파일인 `jest.setup.ts` 파일을 root 디렉토리에 추가해 놓았습니다. 현재는 `Transaction` 데코레이터가 있어도 테스트가 정상적으로 수행되도록 다음과 같은 코드를 추가해놓았습니다.

```ts
jest.mock('typeorm-transactional', () => ({
  Transactional: () => {
    return (target, propertyKey, descriptor) => {
      return descriptor;
    };
  },
}));
```

## 😭 **어려웠던 점**

### mapToHexConverter

첫 시도로는 제네릭을 이용해 return 타입을 넣어주려고 했습니다. 

```ts
const mapToHexColor = <
  T extends { color: THexColor },
  K = Omit<T, 'color'> & { color: string },
>(
  obj: K | K[],
): T | T[] => {
  ...
};

// 사용
const tmp = mapToHexColor<SomeDto>([ ... ]); // 인자가 배열일 경우 tmp 의 타입은 SomeDto[]
const tmp2 = mapToHexColor<SomeDto>( ... ); // 인자가 단일 객체인 경우 tmp2의 타입은 SomeDto
```

그런데 아무리 해도 사용하면 타입이 `SomeDto | SomeDto[]` 가 되었습니다. 이 문제를 해결하기 위해 고군분투했지만 스스로 해결하지 못하고 동료들의 도움을 받아 해결할 수 있었습니다. 제네릭에 넘겨주는 타입이 Return에 대한 타입이 아닌 인자에 대한 타입을 넘겨주고, 생각보다 많은 4개의 제네릭 타입을 이용하여 구현할 수 있었습니다. 이 과정에서 제네릭에 '='이 있으면 생략할 수 있는 제네릭이라는 사실도 알게되었습니다. 

### HexColor 타입

처음에는 HexColor Type을 아래와 같이 만들려고 했습니다. 

```ts
type HexDigit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | 'a' | 'b' | 'c' | 'd' | 'e'| 'f' | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' ;
type HexColor = `#${HexDigit}${HexDigit}${HexDigit}${HexDigit}${HexDigit}${HexDigit}`;
```
하지만 위 코드는 에러가 발생합니다. 왜냐하면 타입스크립트에서 union 타입을 사용할 때 가능한 최대 후보 개수는 100,000개라고 합니다. 위의 코드는 `HexDigit`가 총 22개이므로 `HexColor`는 22^6 = 113,379,904 개가 됩니다. 100,000개는 아득히 뛰어넘기 때문에 위와 같은 타입은 정의할 수 없다고 합니다. 그래서 개선한 방법이 아래와 같은 코드였습니다. ([참고](https://stackoverflow.com/questions/68766792/how-to-create-standalone-type-for-a-hex-color-string))

```ts
type HexDigit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | 'a' | 'b' | 'c' | 'd' | 'e'| 'f' | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' ;
type HexColor<T extends string> =
  T extends `#${HexDigit}${HexDigit}${HexDigit}${infer Rest1}`
    ? Rest1 extends `${HexDigit}${HexDigit}${HexDigit}`
      ? T
      : never
    : never;
type THexColor = HexColor<string>;
```
와 같이 만들었습니다. 하지만 이마저도 제가 원하는 형태는 아니었습니다. 
```ts
const tmp = '#123456'; // 제가 원하는 형태. 하지만 Type 'string' is not assignable to type 'never'. 에러가 발생합니다.

// 올바른 사용방법. 하지만 제가 원하지 않는 형태입니다.
function hex<T extends string>(s: HexColor<T>): T {
    return s;
}
const tmp2 = hex('#ffffff'); // 에러가 발생하지는 않지만 tmp2의 타입은 "#ffffff"가 됩니다. 값이 아니라 타입이 말이죠
```

여러모로 시도한 끝에 원하는 형태의 타입은 만들 수 없다고 판단하여 아래와 같은 타입이 되었습니다.
```ts
type THexColor = `#${string}`;
```

## 🌄 **스크린샷**

![image](https://user-images.githubusercontent.com/32933980/228754576-369e7bae-2167-412f-a5fe-67eac58f377b.png)

jintak0401 committed:
feat: Category Repository 각 함수 임시 구현 (#33)

jintak0401 committed:
test: Category Service 테스트 코드 구현 (#33)

jintak0401 committed:
feat: Category Repository 구현 (#34)

jintak0401 committed:
feat: Category Service 구현 (#34)

jintak0401 committed:
feat: 카테고리를 동일한 상태로 바꿀 경우 ConflictException (#34)

jintak0401 committed:
test: 테스트 코드 오류 수정 (#33)

- stubCategory가 배열이므로 Object.assign 제거
- deleteCategory의 mockResolvedValue를 true로 변경

jintak0401 committed:
test: Category Service 테스트코드 추가 (#33)

- 다른 color와 같은 이름으로 카테고리 수정
- 같은 color와 다른 이름으로 카테고리 수정

jintak0401 committed:
feat: Entity에서 color를 string으로 변경 (#45)

jintak0401 committed:
feat: DTO에서 color를 string으로 변경

jintak0401 committed:
feat: '111111' 색상을 '#111111'로 바꿔주는 유틸 함수 구현 (#45)

jintak0401 committed:
feat: category 관련 타입에서 색상관련 타입 추가 및 변경 (#45)

jintak0401 committed:
feat: 각 함수의 return 값에 mapToHexColor 적용 (#45)

jintak0401 committed:
Merge branch 'dev' into feat/category_service#34

jintak0401 committed:
refactor: 변경된 경로로 import 수정 (#34)

jintak0401 committed:
feat: THexColor 타입 types/index.ts에 추가 (#34)

jintak0401 committed:
feat: THexColor 타입 적용 (#34)

jintak0401 committed:
fix: updateCategory에서 동일한 이름의 카테고리가 존재하는지 확인하는 로직 추가 (#34)

jintak0401 committed:
test: color stub '111111', '222222'로 변경

jintak0401 committed:
test: Transactional 데코레이터를 위한 jest setup

jintak0401 committed:
test: color가 '11111', '22222'로 변경되어 발생하는 에러 해결

jintak0401 committed:
refactor: CategoryArgs 앞에 'I' 붙이도록 이름 변경 (#34)

jintak0401 committed:
Merge branch 'dev' into feat/category_service#34

seoko97 commented on src/api/category/category.repository.ts:
```diff
@@ -1,7 +1,96 @@
+  IUpdateCategoryArgs,
+} from '@/types/args/category';
+
+const SELECT = {
```
상수 파일로 이동해도 괜찮을 것 같습니다!

jintak0401 replied:
상수이긴 하지만 이 파일 내에서만 사용해서 이 곳에 작성했습니다. 그래도 옮기는게 좋을까요?

jintak0401 replied:
협의 결과 옮기는 것으로 결정되었기 때문에 constants 파일로 `CATEGORY_SELECT` 라는 변수명으로 옮겼습니다.

pyo-sh commented on src/utils/color-converter.ts:
```diff
@@ -0,0 +1,24 @@
+
+const strToHexColor = (color: string): THexColor => `#${color}`;
+
+const mapToHexColor = <
+  IType extends { color: string },
+  TArgs = IType | IType[],
+  TForm = { color: THexColor } & Omit<IType, 'color'>,
+  TReturn = TArgs extends any[] ? TForm[] : TForm,
+>(
+  obj: TArgs,
+): TReturn => {
+  if (obj instanceof Array) {
+    return obj.map(({ color, ...rest }) => ({
+      ...rest,
+      color: strToHexColor(color),
+    })) as TReturn;
+  }
+
+  const { color, ...rest } = obj as unknown as IType;
+  return { ...rest, color: strToHexColor(color) } as TReturn;
+};
```
`mapToHexColor` 타입 에러를 도울 당시 이런 의문도 있었습니다.

```ts
const convertHexColor = <
  Type extends { color: string },
  Form = { color: THexColor } & Omit<Type, 'color'>
>(obj: Type) => {
  const { color, ...rest } = obj;
  return { ...rest, color: strToHexColor(color) } as Form;
}
```

와 같이 함수를 구성하고, 배열에 사용할 땐 고차함수를 사용하는 것을요.
```ts
(await this.categoryRepo.readCategory(userId))).map(mapToHexColor);
```

하지만 완성본을 보니 해당 함수 구현도 깔끔해 보이는 것 같습니다!
정말 제가 뿌듯하네요,,,

jintak0401 replied:
생각해보니 그런 방법도 있었군요. 너무 제 생각에만 빠져있다보니 그런 생각은 못했네요...

pyo-sh commented on src/api/category/category.service.ts:
```diff
@@ -1,41 +1,98 @@
+
+    if (!category) {
+      throw new ConflictException('Category does not exist');
+    } else if (categoryByName) {
+      throw new ConflictException('Category already exists');
+    } else if (
```
카테고리의 이름은 중복될 수 있는것 아닌가요?
예를 들어 사용자가 동일한 카테고리 이름을 생성하되, 이를 분리해서 사용하고 싶은 마음인 것처럼요

jintak0401 replied:
제 생각에는 카테고리 이름은 사용자마다 고유해야할 것 같습니다. 같은 이름으로 카테고리를 생성해서 클라이언트의 왼쪽 사이드바에서 보는 것을 상상해보았는데, 편의성보다는 오히려 더욱 헷갈릴 것 같습니다.

seoko97 commented on src/api/category/category.service.ts:
```diff
@@ -1,41 +1,98 @@
+    updateCategoryArgs.categoryName ??= category.name;
+    updateCategoryArgs.color ??= category.color;
+
+    if (!category) {
```
`updateCategory`에 할당하기 전 먼저 검사를 진행해줘야 되지 않을까요?
해당부분에서 검사를 진행한 이유가 있는지 궁금합니다!

jintak0401 replied:
그렇네요. `category`가 null이 아니라는 것이 확인된 이후에 `updateCategoryArgs`의 `categoryName`과 `color`에 할당해주어야 `category`가 null일 때 에러가 발생하지 않겠네요

jintak0401 replied:
수정하였습니다.

seoko97 commented on src/dto/category/category-create-req.dto.ts:
```diff
@@ -1,19 +1,18 @@
-    minimum: 0x000000,
-    maximum: 0xffffff,
-    default: Buffer.from('0x52d681'),
+    description: '카테고리 색상, #000000 ~ #ffffff',
```
만약 `color`가 `mmmmmm` 등과 같이 hex 값이 아니라면 어떻게 되나요?

pyo-sh replied:
저도 코드에서 Validation 부분은 확인하지 못 했네요!

jintak0401 replied:
밑에 보시면 `@IsHexColor` 라는 데코레이터를 사용하고 있습니다. 여기에서 색상값으로 올바른지 판단하고 있습니다.

jintak0401 replied:
![image](https://user-images.githubusercontent.com/32933980/228768314-169704db-c0bd-408f-81f7-bca09db91ec8.png)

이와 같은 데이터로 swagger에서 테스트를 진행하면

![image](https://user-images.githubusercontent.com/32933980/228768411-1ee54908-b1ee-41a0-94ef-f7c0e25e7d09.png)

이와 같이 에러를 내뱉어줍니다.

seoko97 commented on src/utils/color-converter.ts:
```diff
@@ -0,0 +1,24 @@
+
+const strToHexColor = (color: string): THexColor => `#${color}`;
+
+const mapToHexColor = <
```
고생하셨습니다..

jintak0401 committed:
chore: jest 설정 파일 exclude에 추가

jintak0401 committed:
refactor: Category Repository의 SELECT 객체 constants로 이동 (#34)

jintak0401 committed:
fix: Category Update 버그성 코드 해결 (#34)

category가 존재함을 확인한 이후에 updateCategoryArgs의 categoryName과 color를 category의 것으로 할당하도록 수정

pyo-sh reviewed (approved):
고생하셨습니다! 백엔드의 끝이 보이는군요
